quickwit-storage/src/storage_with_upload_cache.rs

```diff
@@ -0,0 +1,630 @@
+use std::sync::Arc;
+
+use async_trait::async_trait;
+use bytes::Bytes;
+use serde::{Deserialize, Serialize};
+use tokio::sync::Mutex;
+use tracing::warn;
+
+use crate::{LocalFileStorage, PutPayload, Storage, StorageErrorKind, StorageResult};
+
+// TODO(asking): could we make it hidden with `.quickwit-cache`
+const INTERNAL_CACHE_DIR_NAME: &str = "quickwit-cache";
+const CACHE_TEMP_FILE_EXTENSION: &str = "_quickwit_cache_temp_file";
+
+/// Capacity encapsulates the maximum number of items a cache can hold.
+/// We need to account for the number of items as well as the size of each item.
+#[derive(Clone, Copy, Debug, Serialize, Deserialize, PartialEq)]
+pub struct CacheParams {
+    /// Maximum of number of files.
+    max_num_files: usize,
+    /// Maximum size in bytes.
+    max_num_bytes: usize,
+    /// Maximum allowed size of a single file.
+    max_file_size: usize,
+}
+
+impl Default for CacheParams {
+    fn default() -> Self {
+        Self {
+            max_num_files: 1000,
+            max_num_bytes: 10_000_000_000, // 10GB
+            max_file_size: 2_500_000_000,  // 2.5GB
+        }
+    }
+}
+
+/// A storage keeping the most recently uploaded files on local disk.
+///
+/// This storage is meant to be used during indexing and merging where the `put` and `copy_to_file`
+/// methods are the most utilized.
+/// As a result, only the methods `put`, `copy_to_file`, and `delete` interact with the local file
+/// system.
+/// Other operations, such as `get_all` or `get_slice,` directly interact with the underlying remote
+/// storage.
+pub struct StorageWithUploadCache {
+    /// The remote storage.
+    remote_storage: Arc<dyn Storage>,
+    /// The backing local storage.
+    local_storage: LocalFileStorage,
+    // The underlying cache.
+    cache_items: Mutex<HashMap<PathBuf, usize>>,
+    /// The parameters of the cache.
+    cache_params: CacheParams,
+}
+
+impl StorageWithUploadCache {
+    /// Create an instance of [`StorageWithUploadCache`]
+    ///
+    /// It needs both the remote to work with.
+    pub fn create(
+        remote_storage: Arc<dyn Storage>,
+        cache_dir: &Path,
+        cache_params: CacheParams,
+    ) -> StorageResult<Self> {
+        let local_storage_root = cache_dir.join(INTERNAL_CACHE_DIR_NAME);
+        fs::create_dir_all(local_storage_root.as_path())?;
+        let mut file_entries = HashMap::new();
+        let mut total_size_in_bytes: usize = 0;
+        for dir_entry_result in fs::read_dir(local_storage_root.as_path())? {
+            let dir_entry = dir_entry_result?;
+            let path = dir_entry.path();
+            if path.is_file() {
+                let relative_file_path = path
+                    .strip_prefix(local_storage_root.as_path())
+                    .map_err(|err| StorageErrorKind::InternalError.with_error(err))?;
+                let file_size = dir_entry.metadata()?.len() as usize;
+
+                // TODO: check if special temp file and remove it
```

Comment:
here we need to remove files with `CACHE_TEMP_FILE_EXTENSION`?

Reply:
Yes